[WIP] dynamic allocation support for Cook scheduler

solnic opened the pull request.

solnic committed:
Move CoarseCookSchedulerBackend to org.apache.spark.scheduler.cluster.cook

solnic committed:
Add test suite class for CoarseCookSchedulerBackend

  * add a simple `isReady` test to make sure things run and pass

solnic committed:
Add doKillExecutors to cook scheduler

solnic committed:
Remove executor=>job mapping after aborting a job in cook

solnic committed:
Add doRequestTotalExecutors to cook scheduler

Make sure we don't exceed executorLimit in cook scheduler

wyegelwel commented on core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala:
```diff
@@ -182,33 +197,55 @@ private[spark] class CoarseCookSchedulerBackend(
-    loop(maxCores - totalCoresRequested, Nil).reverse
+      if (jobs.size == executorLimit) jobs
+      else
+        if (maxCoresPerJob <= 0) jobs
```
if `maxCoresPerJob` is `<= 0` seems like invalid input.

solnic replied:
This is how it was done originally. It's taken from spark's config and defaults to `5`. I'm not entirely sure why it's like that here :) Maybe for some testing purposes, hard to tell.

wyegelwel commented on core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala:
```diff
@@ -182,33 +197,55 @@ private[spark] class CoarseCookSchedulerBackend(
-    runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
-    jobClient.submit(jobs.asJava, jobListener)
+
+    if (!jobs.isEmpty)
```
I think this if is unnecessary as sum will return 0 if the list is empty, and ++ will concat a list with an empty list. What was your concern here?

solnic replied:
Test didn't pass without this conditional. I'll try to simplify it.

wyegelwel replied:
What failed?

On Wed, May 25, 2016, 6:16 PM Piotr Solnica notifications@github.com
wrote:

> In
> core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala
> https://github.com/mojotech/spark/pull/1#discussion_r64663245:
> 
> >    }
> > 
> >    def requestRemainingCores() : Unit = {
> >      val jobs = createRemainingJobs()
> > -    totalCoresRequested += jobs.map(_.getCpus.toInt).sum
> > -    runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
> > -    jobClient.submit(jobs.asJava, jobListener)
> >   +
> > -    if (!jobs.isEmpty)
> 
> Test didn't pass without this conditional. I'll try to simplify it.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/mojotech/spark/pull/1/files/d7d3110f1b513f75b9f86c9f785f443332e63c63#r64663245

wyegelwel commented on core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala:
```diff
@@ -115,6 +118,11 @@ private[spark] class CoarseCookSchedulerBackend(
   }
 
   var runningJobUUIDs = Set[UUID]()
+  var executorsToJobs = HashMap[String, UUID]()
```
Can you rename this to `executorsToJobIds`? I had to scroll up here to ensure that it was string -> UUID.

wyegelwel commented on core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala:
```diff
@@ -182,33 +197,55 @@ private[spark] class CoarseCookSchedulerBackend(
+
+  override def doKillExecutors(executorIds: Seq[String]): Boolean = {
+    val uuids = executorIds.map(x => executorsToJobs(x))
+    jobClient.abort(uuids.asJavaCollection)
```
What happens if the abort fails?

solnic replied:
I didn't cover this yet since I wasn't sure how to simulate a crash. Would it be OK if I refactor this a bit so that I can mock the jobClient in tests? Then I could add proper handling of errors, it's important to make sure that the executorsToJobs is updated only when we actually succeeded with aborting jobs.

solnic committed:
Remove redundant conditional

solnic committed:
Rename executorsToJobs=>executorsToJobIds

solnic committed:
Handle JobClientException when trying to kill a job

solnic committed:
Use val and remove sparkConf placeholder var

solnic committed:
Make sure totalFailures is 0 when jobs were aborted

solnic committed:
Track aborted jobs and make sure we decrease num of requested cores

solnic committed:
Make sure abortedJobIds is being updated on job status change

solnic committed:
Extend test with scaling executors up

solnic committed:
Use more jobs/cores in the cook test

solnic committed:
Calculate current limit of cores when requesting remaining cores in dyn-alloc mode

solnic committed:
Remove unused executorLimit method

solnic committed:
Remove executor-id=>job-id map

Turned out when cook is used executor ids passed to doKillExecutors are
actually job UUIDs, so we can simply use those for aborting jobs in Cook

solnic committed:
Request remaining cores in doRequestTotalExecutors

This turned out to be needed in cook's scheduler as it works differently
in dyn-alloc mode than mesos scheduler. Now after dyn-alloc kills jobs,
we can be sure that new executors (jobs) will be created when a new
spark task needs to be run.

solnic committed:
Simplify test for the initial cook scheduler state

solnic committed:
Make sure abortedJobIds contain correct job UUIDs

solnic commented:
closing in favor of #2
